### PR TITLE
distsql: use a more accurate type of the context of distsql

### LINKED ISSUE #51617: Use a more accurate `DistSQLContext` to avoid depending on the whole session variables
## Enhancement

As the `DistSQLContext` only uses a small subset of all variables and session tools, we can use a more accurate `DistSQLContext` to avoid depending on a whole session variables. And it'll help us to understand how to clone it.

There are also problems about the `Reset` and dependency loop. Obviously the `DistSQLContext` should be re-generated after each statement, but some fields can only be defined well in session (because of circular dependency).

### PR DESCRIPTION
**Pull request by YangKeao**
### What problem does this PR solve?

Issue Number: close #51617

Problem Summary:

The `distsql` only needs a subset of the whole session variables and context, so we can use a more accurate type to represent it. This refactor will help us to clone it in the future.

### What changed and how does it work?

1. Give a struct `DistSQLContext`, which contains all information needed by the session.
2. Add a cache on the `StmtCtx`, because the cache should be invalidated after each execution.
3. Remove the dependency of `stmtctx` (and session, maybe) from `distsql`.

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [x] No need to test
  > - [ ] I checked and no code files have been changed.
  > It should be covered by existing tests. Some tests also need to be modified.

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: pkg/executor/adapter.go [hunk 1]
```diff
@@ -504,21 +505,25 @@ func (a *ExecStmt) Exec(ctx context.Context) (_ sqlexec.RecordSet, err error) {
...
 		_, planDigest := GetPlanDigest(stmtCtx)
-		_, digest := stmtCtx.SQLDigest()
-		stmtCtx.RunawayChecker = domain.GetDomain(sctx).RunawayManager().DeriveChecker(sctx.GetSessionVars().StmtCtx.ResourceGroupName, stmtCtx.OriginalSQL, digest.String(), planDigest.String())
+		_, sqlDigest := stmtCtx.SQLDigest()
+		stmtCtx.RunawayChecker = domain.GetDomain(sctx).RunawayManager().DeriveChecker(sctx.GetSessionVars().StmtCtx.ResourceGroupName, stmtCtx.OriginalSQL, sqlDigest.String(), planDigest.String())
 		if err := stmtCtx.RunawayChecker.BeforeExecutor(); err != nil {
 			return nil, err
 		}
 	}
+	a.initKvExecCounter()
```
Comment (YangKeao) [inline pkg/executor/adapter.go:519]: The most risky part of this PR. I adjusted the order of initialization of multiple fields.
1. The `RunawayChecker`
2. The `KvExecCounter`
3. The plan from `Execute` plan

The `plan` should be initialized before `KvExecCounter` and `RunawayChecker`, to use the correct digest. However, previously the `plan` for `Execute` statements are extracted inside `buildExecutor` (from `ExecuteExec`). I modifed to build the `plan` from `*plannercore.Execute`, and initialize them before `buildExecutor`.

I hope everything needed by executors can be initialized before `buildExecutor`. However, it's not true now, a lot of fields are initialized after `buildExecutor` (by setting fields of `StmtCtx`), and the dependency between them are not clear.

**Review by lcwangchao — APPROVED** [review]

File: pkg/sessionctx/stmtctx/stmtctx.go [hunk 2]
```diff
@@ -167,6 +168,15 @@ type StatementContext struct {
 	// errCtx is used to indicate how to handle the errors
 	errCtx errctx.Context
 
+	DistSQLContext distsqlctx.DistSQLContext
```
Comment (zimulala) [inline pkg/sessionctx/stmtctx/stmtctx.go:171]: Do we use this variable？

**Review by zimulala — APPROVED** [review]

File: pkg/sessionctx/stmtctx/stmtctx.go (hunk 2 above)
Comment (YangKeao) [reply pkg/sessionctx/stmtctx/stmtctx.go]: No. I'll delete it :beers: .

**Review by XuHuaiyu — APPROVED** [review]

**Pull request merged**